### PR TITLE
Add tests for module specifications

### PR DESCRIPTION
**Pull request by rjmholt**
## PR Summary

Testing to validate https://github.com/PowerShell/PowerShell/pull/7126. (This should be merged first)

Adds testing specifically for all the code around ModuleSpecification objects. In particular:
* The `ToString()` funtionality (paired with parsing back to the object)
* The equality comparisons
* Hashcode logic

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [x] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

File: test/powershell/engine/Module/ModuleSpecification.Tests.ps1 [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+        $ht2 = $ht.Clone()
+        $ht2.$Key = $Value
+        $ht2
+    }
+}
+
+# Take a base hashtable and produce all possible
+# combinations of that hashtable with the keys
+# and values in the key/value hashtable
+function PermuteHashtable
```
Comment (iSazonov) [inline test/powershell/engine/Module/ModuleSpecification.Tests.ps1]: Can the test structures be initialized statically? It seems to be more simple and visual than to understand this helper code. What do you think?

Comment (rjmholt) [reply test/powershell/engine/Module/ModuleSpecification.Tests.ps1]: I've been thinking about this. You're definitely right that the number of possible ModuleSpecification field combinations is lower than I imagined initially.

On the other hand, for every field that could be added to a ModuleSpecification, we would have to generate exponentially more tests vs add a single field in here. Perhaps that's unlikely. But I can see something like PSEdition being added, which would take the number of static cases from 6 to 12.

**Review by iSazonov — APPROVED** [review]

File: test/powershell/engine/Module/ModuleSpecification.Tests.ps1 [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+# Take a hashtable and return two copies,
+# one with the key/value pair, the other without
+function PermuteHashtableOnProperty
+{
+    param([hashtable[]]$Hashtable, [string]$Key, [object]$Value)
+
+    foreach ($ht in $Hashtable)
+    {
+        $ht.Clone()
+    }
```
Comment (BrucePay) [inline test/powershell/engine/Module/ModuleSpecification.Tests.ps1]: Just as an aside doing `@{} + $ht` is essentially equivalent to `.Clone()`

File: test/powershell/engine/Module/ModuleSpecification.Tests.ps1 [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+                $clone.Guid | Should -Be $ModuleSpecification.Guid
+            }
+        }
+    }
+
+    Context "ModuleSpecification comparison" {
+
+        BeforeAll {
+            $modSpecAsm = [Microsoft.PowerShell.Commands.ModuleSpecification].Assembly
+            $modSpecComparerType = $modSpecAsm.GetType("Microsoft.PowerShell.Commands.ModuleSpecificationComparer")
```
Comment (BrucePay) [inline test/powershell/engine/Module/ModuleSpecification.Tests.ps1]: You could use `using namespace` to get rid of the long type names.

**Review by BrucePay — APPROVED** [review]

File: test/powershell/engine/Module/ModuleSpecification.Tests.ps1 (hunk 2 above)
Comment (rjmholt) [reply test/powershell/engine/Module/ModuleSpecification.Tests.ps1]: Aha! Which do you consider nicer (for our purposes)?

**Comment by iSazonov** [comment]
@adityapatwardhan Seems the PR is ready to merge.

**Pull request merged**